Build responsive side navigation drawer

bartaz opened the pull request:
## Done

Adds responsive styles for side navigation.
See **Tasks** below for details on specific areas touched.

Fixes #2934

## QA

Run `./run` or [demo](https://vanilla-framework-canonical-web-and-design-pr-2977.run.demo.haus/docs/examples/templates/maas-docs-grid)

Check steps below on the example template page:

- load docs on large screen, sidenav should be in the grid
- resize to small screen, sidenav should hide
- toggle sidenav open - it should open
- toggle it back
- resize screen back to large - snap back to grid
- resize to small, open side nav, resize to large with open side nav
- make screen small (height and width) so sidenav needs to be scrollable, check if that works
- QA small screen on real phone


## Screenshots

<img width="544" alt="Screenshot 2020-04-15 at 15 02 39" src="https://user-images.githubusercontent.com/83575/79340159-32736a00-7f2a-11ea-8b64-359a1612d13b.png">




### Tasks

- [x] move colors to themes
- [x] dark theme
  - what to do with buttons? (buttons don't have themes)
- [x] buttons (use p-button vs own pattern classes)
- [x] icons for < >
- [ ] expand/collapse aria
  - [x] added `aria-controls`
  - [ ] `aria-hidden`/`aria-expanded` turns out problematic because technically we would need to change it on resize
- [x] don't animate when resizing the screen
  - possible but requires a bit of JS to toggle additional class and listen to resize event (see commit https://github.com/canonical-web-and-design/vanilla-framework/pull/2977/commits/54a488f042fa6a24875f56ed7f9cf0f45ba0f43f)
- [x] when sidenav is expanded on small screen, when you resize to bigger screen sidenav goes to content, but overlay stays open
- [ ] overflow hidden on body when drawer is open (?)
   - this will need to react to resize as well to prevent situation when JS adds this to body but doesn't take it away on resize
- [x] close on overlay click

webteam-app commented:
Starting demo at: https://vanilla-framework-canonical-web-and-design-pr-2977.run.demo.haus/

anthonydillon commented:
@bartaz This is looking awesome! great work. I have a few things if they help:
- The Toggle doesn't look like it aligns with the content because the base button is adding some padding. @lyubomir-popov I am not sure minus margin is good here as the button has a hover color so maybe `p-button--neutral` would be better?
![image](https://user-images.githubusercontent.com/1413534/78939790-ac2fd180-7aac-11ea-9f4e-17fb5420a82f.png)

- When crossing the breakpoint from large to small you see a quick flash of animation.
- Clicking on the overlay should close the menu too.

All small things but thought I should mention them.

bartaz commented:
@anthonydillon Thanks for comments. I'm aware of these issues.

> * The Toggle doesn't look like it aligns with the content because the base button is adding some padding. @lyubomir-popov I am not sure minus margin is good here as the button has a hover color so maybe `p-button--neutral` would be better?

I just used buttons cause they were straightforward here. I'm not sure if the design assumed these will be buttons or bigger clickable area.

It would make sense to make whole this area clickable on mobile:
<img width="336" alt="Screenshot 2020-04-10 at 08 17 07" src="https://user-images.githubusercontent.com/83575/78967867-bcbf6680-7b03-11ea-8f99-7ca6e6128960.png">

Neutral button works in content (as border aligns):
<img width="255" alt="Screenshot 2020-04-10 at 08 16 27" src="https://user-images.githubusercontent.com/83575/78967890-cd6fdc80-7b03-11ea-8cab-53a3105f40bd.png">

but on the sidenav it feels busy with all the borders:

<img width="347" alt="Screenshot 2020-04-10 at 08 16 33" src="https://user-images.githubusercontent.com/83575/78967910-d6f94480-7b03-11ea-8c44-28e9118b9699.png">

I guess the content button could be neutral (to keep border) and sidnav close button can be base (to hide borders).

We can adjust these on design review level.

But also, because they are simply a buttons, we can just put them in the example code, but they are external patterns. So project will be free to use any other button anyway.
Unless we want to codify it in the pattern, using custom class names (not `p-button--`) to make sure these styles are always controlled by side nav.

I'm not sure if that's something we've been doing in Vanilla before. I guess other patterns with buttons just use `p-button`.

> * When crossing the breakpoint from large to small you see a quick flash of animation.

Yes, I noticed that, but there is no easy way around that.
We want sidenav hiding / showing animated. So it animates when hiding 'automatically' on resizing as well.
To prevent it we would need to have additional class name for hiding with animation. And JS for toggling would need to be aware of these 3 states of just two.

I'll experiment, but I wouldn't like to overcomplicate JS to avoid that flicker.

> * Clicking on the overlay should close the menu too.

Of course.

anthonydillon commented:
I would like to see the buttons codified into the pattern to give us consistent styling across projects and a clean upgrade path for the component.

Yes, agree we should not over complicated the js here just worth an issue to think of elegant solutions.

bartaz commented:
> I would like to see the buttons codified into the pattern to give us consistent styling across projects and a clean upgrade path for the component.

@anthonydillon Agreed. Having a quick look it seems like that would require a dependancy on `_patterns_button.scss` - because that's where different flavors of buttons are defined.
But they are all defined as mixins, so maybe it's a good time to extract those button mixins to some shared global/base space? So any component can consistently reuse button mixins without relying on `p-button` class names?

anthonydillon commented:
Do you mean to achieve something like this?
p-side-navigation
```scss
@mixin side-navigation {
  @import "p-buttons";

  .p-side-navigation__toggle {
    @extend vf-base-button;
    ....
  }
}
```

bartaz commented:
Yes, but to avoid pattern dependency do it as:

```scss
@mixin side-navigation {
  @import "base-button-definitions";

  .p-side-navigation__toggle {
    @extend vf-base-button;
    ....
  }
}
```

anthonydillon commented:
Let's catch up on Tuesday. I would like to keep the styles for buttons in the buttons file. Otherwise it becomes difficult to maintain and for other developers to know where the styles are defined. Let's see if we can define the styles within the buttons file but make them available globally. This could be a change to the method of inclusion.

lyubomir-popov commented:
@bartaz can we use the base button (no border)? Also can we apply a compensation margin so the icon, not the button edge, aligns with left edge of things above/under?

bartaz commented:
@lyubomir-popov We can, but should we?

I think @anthonydillon specifically mentioned using base button (with border) to avoid compensating with negative margin.

Negative margin makes the button stick out on hover:


<img width="392" alt="Screenshot 2020-04-14 at 16 06 51" src="https://user-images.githubusercontent.com/83575/79234287-57080d00-7e6a-11ea-999a-a9b5ab1dcdff.png">

We shouldn't override styling with negative margins and other overrides. If given style of the button doesn't work in given context we should use different button style.
Or - if any button style doesn't work for us - maybe it shouldn't be a button?

lyubomir-popov commented:
@bartaz you're right. I would vote for note using a button. On mobile, the latest requested change floats the icon to the right, and the whole area should be clickable.

To indicate a hover, the icon colour could change from #666 to #111: 
![image](https://user-images.githubusercontent.com/2741678/79235872-14433680-7e64-11ea-9c94-3d76c7d5639f.png)

What do you think @spencerbygraves ?

bartaz commented:
> @bartaz you're right. I would vote for note using a button. On mobile, the latest requested change floats the icon to the right, and the whole area should be clickable.

@lyubomir-popov @spencerbygraves 

Technically it should still be a button (from HTML point of view). Of course, we can remove all default styling from it and pretend it's not a button.
But, why can't we use one of a handful of existing button styles? Do we need to invent new styles for a 'clickable thing' just for this toggle? It is that different from other buttons we have on the page?
On hand we want to keep things consistent across patterns and project, do we need to invent new styles here?

If this toggle shouldn't look like a button - what about the other one (on the top of open side navigation menu) - I suppose it should have the same style?

webteam-app commented:
Starting demo at: https://vanilla-framework-canonical-web-and-design-pr-2977.run.demo.haus/

lyubomir-popov commented:
> But, why can't we use one of a handful of existing button styles?

If we decide the alternatives are too costly, I can live with what is currently implemented.
I think a base button (no outline, icon visibly misaligned) is a lesser compromise than a bordered button.

I also think this is a problem we've come across before, in different contexts so maybe we can use this case as an opportunity to discuss it in a little more detail.

1. For example, if you look at the snapcraft Products menu and User name dropdown, we have a similar misalignment. Both buttons include padding, which is proportional to the button size, and both are offset from the edge of the fixed width container by different amounts:
![image](https://user-images.githubusercontent.com/2741678/79348299-baf30a00-7f2c-11ea-9884-a31957b23621.png)

2. Jaas.ai/models - similar problem, button and label not aligned:
![image](https://user-images.githubusercontent.com/2741678/79348479-f4c41080-7f2c-11ea-955a-825273de68c3.png)

One way to fix these would be to have a modifier button/menu/select etc that doesn't include a hover background, and thus doesn't require left/right padding.

anthonydillon commented on templates/docs/examples/templates/maas-docs-grid.html:
```diff
@@ -52,113 +31,119 @@
-          </li>
-        </ul>
+        <button class="p-side-navigation__toggle js-drawer-toggle" aria-controls="drawer">
+          <span>Toggle side navigation</span>
```
Is this span required?

anthonydillon commented on scss/_patterns_side-navigation.scss:
```diff
@@ -1,15 +1,109 @@
+
+  .p-side-navigation__toggle--in-drawer {
+    &::before {
+      transform: rotate(-90deg);
```
I think these transforms are the wrong way around.

anthonydillon commented on scss/_settings_colors.scss:
```diff
@@ -77,6 +79,7 @@ $colors--dark-theme--text-muted: rgba($colors--dark-theme--text-default, 0.55) !
 $colors--dark-theme--background-default: hsl(0, 0%, 15%) !default;
 $colors--dark-theme--background-alt: hsl(0, 0%, 20%) !default;
 $colors--dark-theme--background-highlight: rgba($colors--dark-theme--text-default, 0.05) !default;
+$colors--dark-theme--background-overlay: transparentize($color-dark, 0.15) !default;
```
Should this switched to `$color-light` to add contrast to the dark theme version?

bartaz replied:
@lyubomir-popov ? What should be the color of overlay in dark theme?

lyubomir-popov replied:
In the design, the overlay styling was taken from the RBAC overlay, which is 40% #111 for the light theme.  

Inverting to light overlay seems strange, the background has a high contrast:
![image](https://user-images.githubusercontent.com/2741678/79358624-58543b00-7f39-11ea-8c0e-314aee77aadc.png)

Dark p strip underneath not ideal too at this opacity:
![image](https://user-images.githubusercontent.com/2741678/79356582-10ccaf80-7f37-11ea-9528-ff6306ae1c92.png)

I think dark overlay is better for both, and I much prefer the .85 opacity used by @bartaz. In rbac, the user is supposed to be able make out what is underneath while the overlay is on, but here that's not necessary. So I think what is currently happening on this pr is best:
![image](https://user-images.githubusercontent.com/2741678/79357487-0232c800-7f38-11ea-816d-15b72546c47d.png)

Separately, can we have a box-shadow too, like cards? Then once we have  a proposal for container shadows/elevaiton we can go back and update the shadow depth to something more suitable.

bartaz replied:
@lyubomir-popov What would be the purpose of the box shadow if we have an overlay?

Dark box shadow is completely invisible on such overlay.

It is visible when it shouldn't (when drawer is offscreen it sticks from the left):
<img width="137" alt="Screenshot 2020-04-16 at 07 38 18" src="https://user-images.githubusercontent.com/83575/79418657-765f8100-7fb5-11ea-9f65-4558b5b80e85.png">

We can prevent this from happening of course, but why applying shadow if it will be not visible when drawer is open anyway?

anthonydillon replied:
I did think drop shadow would help but may need to be increased from the normal card shadow. Also, I was thinking it might look nice if we transition the shadow spread. So off canvas has 0 spread were as when the menu is revealed the shadowed spread is set and transitioned to at the same rate as the slide out. This would solve the shadow visible off screen and might make the shadow more visible to the eye.

bartaz replied:
Box shadow will be useful, if it's visible on top of overlay. We would need to adjust the opacity of overlay and shadow to achieve that. With overlay appearing immediately when sidenav starts animating into the screen there is no way to make shadow visible (with our current colors).

Animating shadow is an interesting idea. But I wonder if this will not make it look like drawer is 'elevating' during animation, which could be a bit unexpected.

lyubomir-popov replied:
We recently created a set of rules for different things, one of them was anything that overlaps something else should have a box shadow, so it's a consistency thing. I was wary of suggesting a bespoke drop shadow, but ideally it should be something like the google maps one: 
![image](https://user-images.githubusercontent.com/2741678/79434399-39e85100-7fc6-11ea-9e3f-f1b528fb2403.png) we just have to define our own in the shadow hierarchy piece of work.

I know it won't be very visible here, but in the future we may change the opacity of the overlay. It doesn't hurt to apply the rules consistently.

anthonydillon commented on templates/docs/examples/templates/maas-docs-grid.html:
```diff
@@ -52,113 +31,119 @@
+        <div class="p-side-navigation__drawer" id="drawer">
+          <div class="p-side-navigation__drawer-header">
+            <button class="p-side-navigation__toggle--in-drawer js-drawer-toggle" aria-controls="drawer">
+              <span>Toggle side navigation</span>
```
Is this span required?

bartaz replied:
`p-button has-icon` requires span, and this is based on it.
But because we always have icon on the left I think `<span>` may not be needed. I'll have a look. 

Good catch, thanks.

anthonydillon commented on scss/_patterns_side-navigation.scss:
```diff
@@ -1,15 +1,109 @@
+    padding-bottom: $spv-inner--small;
+    padding-left: $sph-inner--small;
+    padding-top: $spv-inner--small;
+    position: sticky;
```
Nice touch :ok_hand:

lyubomir-popov commented:
@bartaz still concerned about the button behaviours, it is even worse on mobile:

![image](https://user-images.githubusercontent.com/2741678/79359760-e1b83d00-7f3a-11ea-9eab-745eec3229e3.png)

That full width centered button is something optimised for purchase flows, not for things like this.

anthonydillon commented:
The width behaviour on small breakpoints can be modified here as this is a pattern.

bartaz commented:
@lyubomir-popov @anthonydillon 

I updated the PR based on your comments:

- removed unnecessary spans from buttons
- prevented buttons from growing on mobile

I have not added box shadow because it's invisible on dark overlay anyway. More details in comment thread above.

bartaz commented on templates/docs/examples/patterns/side-navigation/_toggle_script.js:
```diff
@@ -0,0 +1,70 @@
+
+  // on resize event remove `is-collapsed` class from drawers to prevent them from animating back in
+  // TODO: this should probably be throttled
+  window.addEventListener('resize', function() {
```
@anthonydillon Are you ok with component script listening on resize?

We need that to prevent sidenav animating on screen resize, but it feels like it's getting into area or 'too much JS' for CSS framework.

To use this pattern this (or equivalent) JS is required. And in most cases it will be copied from here blindly.
And it's not bulletproof as it is now. Resize event should be probably throttled (so it's not called million times). This adds more and more to JS that we don't officially publish or support that is required by pattern to work as expected.

bartaz replied:
@anthonydillon Thanks. It was good call.

I changed transition to animation and it allowed to get rid of resize event.
Moving collapsed/expanded to root element also helped with simplifying JS part.

lyubomir-popov commented:
@bartaz nice work, just a couple of niggles:
- is it possible to make the flashing black overlay appear with a fade?
- the button still worries me, but I'm prepard to accept it asa temporary compromise, then work separately on a button that fits the design  (and maye solves similar issues highlighted above with menus with chevrons etc.) I'll make a separate issue for this.

lyubomir-popov commented:
@bartaz I've created a working group [issue ](https://github.com/canonical-web-and-design/vanilla-framework/issues/2982) to discuss how we can achieve the properly aligned version of the toggle.

anthonydillon commented on templates/docs/examples/patterns/side-navigation/_icons.html:
```diff
@@ -1,95 +1,112 @@
+
+  {% if is_dark %}
+  <!-- dark background color should be changed via color theme variables, inline style is used here for example purposes -->
+  {% endif %}
```
Do we need this?

bartaz replied:
I wanted to make it clear that setting background inline is not recommended way. Just in case someone uses example as a reference.

I can remove it. Especially that this example is not exposed in CodePen embeds, it's just listed on Examples pages and screenshoted for Percy.

anthonydillon commented:
The demo's toggle doesn't seem to work on Google Chrome on Android?

bartaz commented on scss/_patterns_side-navigation.scss:
```diff
@@ -1,15 +1,145 @@
+    width: 100%;
+    z-index: 12;
+
+    [class*='p-side-navigation'].is-expanded & {
```
We already have `.p-side-navigation` and `.p-side-navigation--icons` to cover.
It's not much, we can explicitly list them both here. But this just adds when another variants are added.

It really is limiting that we can't rely on main class `p-side-navigation` being ALWAYS on the parent.

anthonydillon replied:
Oh I see, they are both under the same mixin. Normally the variants are there own mixin with shared placeholders.

bartaz replied:
Oh, I haven't notice that. I guess not all patterns do it consistently.

bartaz commented on scss/_patterns_side-navigation.scss:
```diff
@@ -1,15 +1,145 @@
+      transform: translateX(0);
+    }
+
+    [class*='p-side-navigation'].is-collapsed & {
```
We can't if we want to prevent unexpected animations.

Having 'default' (collapsed) state as one without class means that close animation will trigger on page load for example.

To have animation only triggered by user intent, we need to have both of these classes.

bartaz commented:
> The demo's toggle doesn't seem to work on Google Chrome on Android?

@anthonydillon Any specific URL. I tested the MAAS docs example template and seems to be working fine on Android Chrome.

![Screenshot_20200416-163103](https://user-images.githubusercontent.com/83575/79469008-f0b4f300-7fff-11ea-8f0d-93790133a08c.jpg)

anthonydillon commented:
Yeah, this URL: https://vanilla-framework-canonical-web-and-design-pr-2977.run.demo.haus/docs/examples/templates/maas-docs-grid

bartaz commented:
@anthonydillon Hmmm... Works for me. And I think we both have OnePlus with recent Chrome?

anthonydillon commented:
Checking browserstack, working on FF on my phone.

anthonydillon commented:
Works on BrowserStack so must be cache or something, nevermind.

anthonydillon commented:
The double navigations in the docs are causing issues. We should only have a single example per example (if that makes sense):
![image](https://user-images.githubusercontent.com/1413534/79560813-30cbb280-80a0-11ea-954b-7132325b62ff.png)

anthonydillon commented on templates/docs/patterns/navigation.md:
```diff
@@ -89,6 +89,9 @@ To add icons on the left side of the items in side navigation use the `.p-side-n
 View example of the side navigation pattern with icons
 </a>
 
+On small screens side navigation is rendered off-screen as a drawer. To open the side navigation drawer, add `is-expanded` class to
```
Can we title this section JS functionality or something?
